nucleus/security/core/src/main/java/com/sun/enterprise/security/auth/realm/certificate/CertificateRealm.java

```diff
@@ -175,6 +213,67 @@ public String authenticate(Subject subject, X500Principal principal) {
         return principalName;
     }
 
+    private void validateSubjectViaAPI(Subject subject, X500Principal principal,
+            X509Certificate certificate, String componentId) throws LoginException {
+        List<ClientCertificateValidator> validatorList = Collections.emptyList();
+        try {
+            validatorList = loadValidatorClasses();
+        } catch (Throwable exc) {
+            _logger.log(Level.WARNING, "Exception while loading certificate validation class", exc);
+            clientCertificateValidatorMap.remove(Utility.getClassLoader());
+        }
+
+        boolean failed = false;
+        if (!validatorList.isEmpty() && componentId != null) {
+            try (Context context = ctxUtil.get().fromComponentId(componentId).pushContext()) {
+                failed = validatorList.stream().anyMatch(validator
+                        -> !inject(validator, context).isValid(subject, principal, certificate));
+            } catch (Exception exc) {
+                _logger.log(Level.WARNING, "Exception from certificate validation class", exc);
+                clientCertificateValidatorMap.remove(Utility.getClassLoader());
```

Comment:
I think it is safer to say the validation failed when there was an Exception.

`failed =true;`

Reply:
I disagree. For example, if some downstream service were to go down, you don't want to bring the whole authentication down with it, potentially bringing down the whole system.

It's better to default to old behavior if the call fails completely

Reply:
It depends, it could lead to an accepted certificate which should not be allowed access. Hence you are opening up a security vulnerability.

But we can discuss later

Reply:
In that case the API needs to change. `isValid()` needs to return `void` and exception needs to be documented / thrown when authorization fails.
I have no preference which way to do it however, if `ieValid()` returns `boolean` exception means needs to mean service failure and authorization success so not to bring down the system